---
title: "Positive selection scan on Capybara genome"
author: "Santiago Herrera-Alvarez"
date: "September 9, 2018"
output:
  html_document:
    df_print: paged
  pdf_document:
    highlight: kate
geometry: margin=2cm
fontsize: 10pt
---

```{r knitr, echo=FALSE}
knitr::opts_chunk$set(
  comment   = "#",
  results   = "hold",
  # collapse  = TRUE,
  fig.align = "center",
  cache=FALSE,
  message = FALSE,
  warning = FALSE,
  tidy= TRUE,
  tidy.opts=list(width.cutoff=80))
```

```{r}
library(ggplot2)
library(biomaRt)
library(ggrepel)
library(MCMCTreeR)
library(ggtree)
```

## Data: LRTs and PDIs
Amino acid sites within a protein may experience different selective pressures and have different underlying $d_N/d_S$ ratios (Yang et al., 2000). We fitted branch-site models to identify positive selection in a set of 4,452 SCOs obtained using the OrthoVenn software (Wang et al., 2015) on capybara versus four closely related rodents with published whole genome sequences: naked mole rat, degu, chinchilla and guinea pig. **NOTE: results of LRTs and PDIs for each of the 4,452 SCOs are available in the table `LRs-PDI`. The p-values correspond to each LRT**

+ **LRTs**: Likelihood-ratio tests (LRT) between two models of codon evolution were used as a measure of model fit: a branch-site model (model A) in which attempts to detect positive selection acting on a few sites on specified lineages or 'foreground branches' ($d_N/d_S$>1), versus the null model (A null) in which codons can only evolve neutrally ($d_N/d_S$ =1) or under purifying selection (0< $d_N/d_S$ <1; Zhang et al., 2005).

+ **PDI**: ((Capybara, guinea pig), degu) protein trios where used to estimate the difference of the raw p-distance in protein sequence between capybara and (pdCD), and the raw p-distance between guinea pig and degu (pdGD). These two distances were used to calculate the ???protein distance index??? or PDI for each (PDIi = pdCDi ??? pdGDi) which takes a value of zero if the two p-distances are equal, meaning that the capybara and guinea pig lineages have evolved at equal rates since their divergence from their most recent common ancestor (MRCA). A positive value is indicative of accelerated protein evolution in the capybara lineage.

```{r}
tr <- readMCMCTree("../Rodent_phylogenomic_MCMCtree.tre")
g <- groupOTU(tr$apePhy,.node=c("Capybara","Cavia","Degu"))
p2 <- ggtree(g, aes(color=group)) + geom_tiplab() + geom_hilight(30,"steelblue",alpha = 0.6) + xlim(0, 1) + scale_color_manual(values=c("black","orange"))
flip(p2,30,20)

```

**ORANGE:** Scope of the PDI analysis
**BLUE:** Scope of the branch-site models of positive selection

### Exploratory analysis

**NOTE: The file `LRs-dists.csv` contains the following information: SCO-cluster name, LRT value, p-value of LRT, Tamura-Nei distance for DNA seqs, protein raw-p distance, Guinea pig ensembl prot id, gene name**

Tamura-Nei distances of DNA sequences and raw-p distances were computed on ((Capybara, guinea pig), degu) orthologs.

```{r}
t <- read.csv("../LRs-dists.csv",h=TRUE)

#Quick and exploratory plot of Tamura-Nei distance vs, raw-p distance. Note the tight positive correlation between both measures.
d <- t$DNA_tamura.nei.D
d <- as.numeric(levels(d))[d]
t$D_TN <- d
dp <- t$prot_pdistance
dp <- as.numeric(levels(dp))[dp]
t$D_prot <- dp
t2 <- t[complete.cases(t), ]
t2 <- t2[!(t2$LR<0),]

ggplot(t2, aes(x=D_TN, y=D_prot)) + geom_point() + geom_smooth() + xlab("Tamura-Nei distance") + ylab("Protein raw-p distance")
```

## Plot of the genes on a two-dimension graph with coordinates x=LR, y=PDI

We considered genes in the top 5% of the distribution in both dimensions (i.e., the LRTs of branch-site models and the PDI statistic) as positively selected in the capybara lineage.

```{r}
#Extract the genes on the top 5% of the distribution of both axes
labels=t2[t2$LR>=quantile(t2$LR,0.95) & t2$D_prot>=quantile(t2$D_prot,0.95),]

#Labels for genes related to growth and cancer
labels_growth <- t2[t2$gene_name=="PLAGL2" | t2$gene_name=="RFX6",]
labels_cancer <- t2[t2$gene_name=="PARP2",] 
label_insulin <- t2[t2$gene_name=="Ins",]


ggplot(t2, aes(LR, D_prot)) + geom_point(colour="grey87") + geom_point(data=labels, aes(x=LR, y=D_prot), colour="red") + geom_point(data=labels_growth, aes(x=LR, y=D_prot), colour="darkorange1",size=3) + geom_point(data=labels_cancer, aes(x=LR, y=D_prot), colour="darkorange1",size=3) + geom_point(data=label_insulin, aes(x=LR,y=D_prot),colour = "darkorange1", size=3)+ geom_text_repel(data=rbind(labels_growth,labels_cancer,label_insulin),aes(label=gene_name),fontface = 'bold', color = 'black', box.padding = unit(0.35, "lines"), point.padding = unit(0.5, "lines")) + labs(x="LRT", y = "PDI") + theme(panel.background = element_rect(fill = "white", colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(face="bold", size=13),axis.text.y = element_text(face="bold", size=13), axis.title.y = element_text(face="bold", size=15),axis.title.x = element_text(face="bold", size=15)) + theme(legend.position = c(0.75, 0.6),legend.background = element_rect(color = "black", fill = "white", size = 1, linetype = "solid")) + geom_vline(xintercept = quantile(t2$LR,0.95), linetype="dotted") + geom_hline(yintercept = quantile(t2$D_prot,0.95), linetype="dotted")
```

Interestingly, Insulin is not positevely selected in the capybara, although it has been shown to have signatures of positive selection in caviomorph rodents (Opazo et al., 2005). Insulin transcription and secretion mediated by (RFX6 and PDX1) may be part of the underlying mechanism regulating body size evolution in the capybara.

## Fisher's exact test (GO enrichment) of genes under positive selection

```{r}
ensembl=useMart("ensembl")
ensembl_cavia=useMart("ensembl",dataset="cporcellus_gene_ensembl")

all <- t2$ensmbl_CporID
outliers <- labels$ensmbl_CporID

#Extract GO annotations for the 56 genes under positive selection
out_genes <- getBM(attributes=c("ensembl_peptide_id","external_gene_name","go_id","goslim_goa_description"),filters="ensembl_peptide_id",values=as.character(outliers),mart=ensembl_cavia)
#out_names <- unique(out$external_gene_name)

#Extract GO annotations for ALL the 4,452 genes
all_genes <- getBM(attributes=c("ensembl_peptide_id","go_id","goslim_goa_description"),filters="ensembl_peptide_id",values=as.character(all),mart=ensembl_cavia)

#Perform Fisher's exact test
all_GOs <- all_genes$goslim_goa_description
out_GOs <- out_genes$goslim_goa_description
unique_out_GOs <- unique(out_genes$goslim_goa_description)
fisherPvalues_out <- c()
GO_term <- c()

for(i in 1:length(unique_out_GOs)){
	a <- sapply(unique_out_GOs[i],function(x) sum(x==out_GOs))
	b <- sapply(unique_out_GOs[i],function(x) sum(x==all_GOs))
	c <- length(out_GOs)-a
	d <- length(all_GOs)-b
	f <- fisher.test(matrix(c(a,b,c,d),nrow=2,ncol=2,byrow=TRUE),alternative="greater")
  	GO_term <- c(GO_term,as.character(unique_out_GOs[i]))
  	fisherPvalues_out <- c(fisherPvalues_out,f$p.value)
}
fisherPvalues_out_Bonferroni <- p.adjust(fisherPvalues_out,method="bonferroni",n=length(fisherPvalues_out))
out_fisherResults <- data.frame(GO_term,fisherPvalues_out,fisherPvalues_out_Bonferroni)
significativas_out <- out_fisherResults[out_fisherResults$fisherPvalues_out_Bonferroni<0.05,]

significativas_out
```















